### DISCUSSION TITLE
Calculating Dawn, Dusk and Solar Noon

### DISCUSSION THREAD

**Discussion by offsky** [cosinekitty/astronomy#122] [Q&A] [answered]
I am trying to use this library to calculate the times for dawn, dusk and solar noon. It looks like `Astronomy.SearchHourAngle` is the function that I should be using.  The definition of civil dawn/dusk is when the Sun's center is 6 degrees below the horizon.  I converted 6 degrees into hours and came up with this code. It produces a close answer, but it differs by 5-10 minutes from two different reputable sources.  The sunrise/sunset times produced by `Astronomy.SearchRiseSet` are dead on when comparing to these other sources, so I must be doing something wrong for dusk/dawn/noon.

I would appreciate it if someone could comment on whether I am doing this correctly.  Thanks!

```
var observer = new Astronomy.Observer(lat, lon, elev);
var noon = Astronomy.SearchHourAngle("Sun",observer,0,date);
var dawn = Astronomy.SearchHourAngle("Sun",observer,17.6,date);
var dusk = Astronomy.SearchHourAngle("Sun",observer,6.4,date);
```

**Comment by cosinekitty** [comment]
Hi @offsky. Your code for finding noon is exactly right, because that is when the Sun has hour angle = 0. For dawn and dusk, what you are looking for is when the center of the Sun has an altitude angle of -6 degrees for the given observer. However, the hour angle is not the same thing as altitude angle.

As an extreme example, imagine you are looking at the "north star" Polaris. It is very close to the north celestial pole, so it is always in the sky for most observers in the northern hemisphere. For such an observer, Polaris never rises or sets. It just moves around in a tiny circle every 24 hours. However, its hour angle goes through the full range 0..24 hours every sidereal day. Sometimes Polaris is directly above the observer's north celestial pole by a tiny angle, meaning Polaris culminates with a 0 hour angle. About 12 hours later, Polaris will be directly below the north celestial pole by the same tiny angle. This is hour angle = 12.

However, Polaris' altitude angle (how far it is above a given observer's horizon) is almost constant, and is approximately equal to the observer's latitude above the equator. So you can see that the hour angle doesn't tell you anything about when (or if) an object rises/sets.

In general, sunrise does not happen 6 sidereal hours before local noon, nor does sunset happen 6 hours afterward. They vary quite a bit depending on the observer's latitude and the season (winter days are shorter, summer days are longer).

So I'm thinking about how to solve your problem. Currently, there is no simple way to calculate civil dawn/dusk using the existing Astronomy Engine functions. You would have to iteratively home in on the time when the Sun's altitude is -6. This is very similar to what the existing function `SearchRiseSet` does. But this would be complicated. I think this functionality should be available in Astronomy Engine. Over the next few days I will add a new function for calculating these values. There will be nuances:

* Sunrise, sunset, moonrise, and moonset are all defined in terms of when the topmost part of the object appears into (or disappears from) view, not when the center of the body crosses the horizon. I will follow your definition regarding the center of the body rather than the topmost part of the body.
* `SearchRiseSet` corrects for atmospheric refraction, but I believe most legal definitions of civil twilight ignore refraction and calculate the altitude angle as if there is no atmospheric distortion.

To help me implement this, can you give me a link to where you find the reputable sources of test data? I will use them to create unit tests for the new function.

**Reply by cosinekitty** [reply]
I will allow passing in the angle below the horizon as a parameter to the new function, because there are [different definitions of twilight](https://www.timeanddate.com/astronomy/different-types-twilight.html): civil, nautical, and astronomical.

**Reply by offsky** [reply]
Thanks for the detailed explanation.  I did a little more reading and I now understand the difference between hour angles and altitude.  I appreciate your willingness to add this functionality to your wonderful library. This will be a great addition!

I was using the [SunCalc](https://github.com/mourner/suncalc) library to get various times related to the sun and moon. It supports a number of important markers (golden hour, sunset start/end, night, nadir, etc) that you may also want to consider.  It bases its calculations on [this article](https://www.aa.quae.nl/en/reken/zonpositie.html). At a point in my project I needed to also have the dates for the solstices and equinoxes, so I went researching and found your library. In comparing the two libraries and also comparing against other sources (see below), I realized that SunCal isn't accurate, so I plan to switch to using your library. The only missing piece is dawn/dusk, which is why I wrote this question :)

Here are the two sources that I was using to compare results. I realize that these may not be "authoritative", but they match each other, so that seemed pretty good to me.  Also, the lat/lon that these sites use may be different from what I plug into Astronomy Engine which may account for some differences.
https://www.timeanddate.com
https://www.gaisma.com/en/
Also, the built in weather iPhone app.

During my researching, I also found the following javascript libraries which calculate the sun's position. Most seem based on Jean Meeus's 'Astronomical Algorithms' book and only the last one directly calculates dawn/dusk. For the other two, I would have to zero in on the answer by looking at the altitude, like you have suggested.
https://github.com/Fabiz/MeeusJs
https://github.com/andrmoel/astronomy-bundle-js
https://github.com/doniseferi/suntimes

**Comment by offsky** [comment]
I naively went off and attempted the "zero in on it" approach that you suggested with the crude code listed below. It seems to work pretty good and gets within 5 minutes of the times listed on the reference websites that I mentioned in my other comment.  I tested San Diego and Seattle for today, winter and summer.  More testing is clearly needed. Not sure why there is a slight 3-5 min difference.
 
```
var sun, pos;
var chunk = 120; //minutes
var angle = -6;
var date = new Date();
var what = "dawn"; //dusk
while(Math.abs(chunk)>0.5) {
	sun = Astronomy.Equator("Sun",date, observer, true, true);
	pos = Astronomy.Horizon(date,observer,sun.ra,sun.dec,"normal");
	if(pos.altitude>angle) {
		if(chunk<0) chunk = -chunk/2;
	} else {
		if(chunk>0) chunk = -chunk/2;
	}
	if(what=="dusk") date.setTime(date.getTime() + (chunk*60*1000));
	else date.setTime(date.getTime() - (chunk*60*1000));
}
console.log(what,date);
```
Note: this is just prototype code. It assumes that the starting time is between dawn and dusk and it has a crude switch between calculating dawn/dusk.

**Reply by offsky** [reply]
Ok, found the anwser to the 3-5 min difference. Its the "normal" parameter in the Horizon function. Change this to false and the dawn/dusk times are dead-on accurate :)  I guess atmospheric refraction is not used for calculating dawn/dusk like you already mentioned. I should have listened ;)

**Reply by cosinekitty** [reply]
About the 3-5 minute difference, can you try changing the call to `Horizon` to look like this?

```
pos = Astronomy.Horizon(date, observer, sun.ra, sun.dec, null);
```

That will turn off the refraction correction, which I strongly suspect will make it more consistent with the usual definitions of civil twilight.

The documentation for the function is [here](https://github.com/cosinekitty/astronomy/tree/master/source/js#Horizon), if you want to read more about the parameters.

**Reply by cosinekitty** [reply]
Haha! I answered before I saw that you already figured it out. Good job!

**Comment by cosinekitty** [comment, marked as answer]
@offsky I just finished work on the new `SearchAltitude` functions in all 4 supported programming languages (C, C#, JavaScript, and Python). It should be both easier to use and more efficient than the binary search for the desired altitude. Here is the [documentation for the JavaScript version](https://github.com/cosinekitty/astronomy/tree/master/source/js#SearchAltitude). Please let me know how it works for you.

**Reply by offsky** [reply]
Wow, you are crazy fast!  I hadn't even had a chance to implement my hacked version and I was only working in one programming language, not four!  I just tested it out and it works perfectly for me. Thanks for making this great library.
